internal/distro/rhel85/stage_options.go

```diff
@@ -306,3 +310,186 @@ func xorrisofsStageOptions(filename string, arch string) *osbuild.XorrisofsStage
 		IsohybridMBR: "/usr/share/syslinux/isohdpfx.bin",
 	}
 }
+
+func grub2StageOptions(pt *disk.PartitionTable, kernelOptions string, kernel *blueprint.KernelCustomization, packages []rpmmd.PackageSpec, uefi bool, legacy string) *osbuild.GRUB2StageOptions {
+	if pt == nil {
+		panic("partition table must be defined for grub2 stage, this is a programming error")
+	}
+	rootPartition := pt.RootPartition()
+	if rootPartition == nil {
+		panic("root partition must be defined for grub2 stage, this is a programming error")
+	}
+
+	stageOptions := osbuild.GRUB2StageOptions{
+		RootFilesystemUUID: uuid.MustParse(rootPartition.Filesystem.UUID),
+		KernelOptions:      kernelOptions,
+		Legacy:             legacy,
+	}
+
+	if uefi {
+		stageOptions.UEFI = &osbuild.GRUB2UEFI{
+			Vendor: "redhat",
+		}
+	}
+
+	if !uefi {
+		stageOptions.Legacy = legacy
+	}
+
+	if kernel != nil {
+		if kernel.Append != "" {
+			stageOptions.KernelOptions += " " + kernel.Append
+		}
+		for _, pkg := range packages {
+			if pkg.Name == kernel.Name {
+				stageOptions.SavedEntry = "ffffffffffffffffffffffffffffffff-" + pkg.Version + "-" + pkg.Release + "." + pkg.Arch
+				break
+			}
+		}
+	}
+
+	return &stageOptions
+}
+
+// sfdiskStageOptions creates the options and devices properties for an
+// org.osbuild.sfdisk stage based on a partition table description
+func sfdiskStageOptions(pt *disk.PartitionTable, device *osbuild.Device) (*osbuild.SfdiskStageOptions, *osbuild.SfdiskStageDevices) {
+	stageDevices := &osbuild.SfdiskStageDevices{
+		Device: *device,
+	}
+
+	partitions := make([]osbuild.Partition, len(pt.Partitions))
+	for idx, p := range pt.Partitions {
+		partitions[idx] = osbuild.Partition{
+			Bootable: p.Bootable,
+			Size:     p.Size,
+			Start:    p.Start,
+			Type:     p.Type,
+			UUID:     p.UUID,
+		}
+	}
+	stageOptions := &osbuild.SfdiskStageOptions{
+		Label:      pt.Type,
+		UUID:       pt.UUID,
+		Partitions: partitions,
+	}
+
+	return stageOptions, stageDevices
+}
+
+// copyFSTreeOptions creates the options, inputs, devices, and mounts properties
+// for an org.osbuild.copy stage for a given source tree using a partition
+// table description to define the mounts
+func copyFSTreeOptions(inputName, inputPipeline string, pt *disk.PartitionTable, device *osbuild.Device) (
+	*osbuild.CopyStageOptions,
+	*osbuild.CopyStageDevices,
+	*osbuild.CopyStageMounts,
+) {
+	// assume loopback device for simplicity since it's the only one currently supported
+	// panic if the conversion fails
+	devOptions, ok := device.Options.(*osbuild.LoopbackDeviceOptions)
+	if !ok {
+		panic("copyStageOptions: failed to convert device options to loopback options")
+	}
+
+	devices := make(map[string]osbuild.Device, len(pt.Partitions))
+	mounts := make(map[string]osbuild.Mount, len(pt.Partitions))
+	for _, p := range pt.Partitions {
```

Comment:
Need to sort that so that root gets mounted first before efi, even if the efi partition before the root partition in the partition table.